Removing exodus io from IBFEInstrumentPanel class

cpuelz opened the pull request:
removing exodus io and replacing probably with output to STLs.

boyceg commented:
Instead of STL, can you just give access to the meshes themselves, so that users can choose any output format?

cpuelz commented:
Oh okay that sounds good. Like have a getMeterMeshes function?

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________
From: Boyce Griffith <notifications@github.com>
Sent: Sunday, August 12, 2018 8:03 PM
To: IBAMR/IBAMR
Cc: Charles Puelz; Author
Subject: Re: [IBAMR/IBAMR] Removing exodus io from IBFEInstruments class (#281)

Instead of STL, can you just give access to the meshes themselves, so that users can choose any output format?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/IBAMR/IBAMR/pull/281#issuecomment-412380896>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ALNyyxFakVWsd6ORJMAMgr17Wg2sNctjks5uQMJNgaJpZM4V5u4F>.

cpuelz commented:
I went back to look at our previous comments in #272 .  I think I know what to do.

cpuelz commented:
@boyceg @drwells I added some functions to get mesh and system members for the meter objects, and removed things having to do with exodus.  what do you think?

drwells commented:
Looks good to me.

I think it would be preferable, though, to return references instead of pointers so that no one is tempted to double-delete the libmesh data structures.

boyceg commented on include/ibamr/IBFEInstrumentPanel.h:
```diff
@@ -100,15 +99,25 @@ class IBFEInstrumentPanel
      */
-    int getInstrumentDumpInterval() const;
-
+    libMesh::SerialMesh* getMeterMesh(const unsigned int jj) const;
```
It looks like the meshes are created only once, and so it should be safe to replace this return type by `libMesh::MeshBase&`.

boyceg replied:
By the way, since this is indexed by the meter number, there needs to be a method for finding out the number of meters (e.g., `getNumberOfMeterMeshes()`). It doesn't look like such a function currently exists in the interface.

boyceg commented on include/ibamr/IBFEInstrumentPanel.h:
```diff
@@ -100,15 +99,25 @@ class IBFEInstrumentPanel
+    /*!
+     * \return get a pointer to the EquationSystem object for jjth meter mesh
+     */
+    libMesh::EquationSystems* getMeterMeshSystem(const unsigned int jj) const;
```
As above, let's change the return type to `libMesh::EquationSystems&`.

Also, let's go ahead and change the method name to `getMeterMeshEquationSystems`, since `System` is a separate concept from `EquationSystems` in libMesh.

boyceg commented on src/IB/IBFEInstrumentPanel.cpp:
```diff
@@ -313,11 +312,10 @@ IBFEInstrumentPanel::IBFEInstrumentPanel(SAMRAI::tbox::Pointer<SAMRAI::tbox::Dat
       d_meter_mesh_names(),
       d_nodeset_IDs_for_meters(),
-      d_instrument_dump_interval(),
+      d_instrument_dump_interval(), 
```
Let's remove this whitespace-only change.

boyceg commented on src/IB/IBFEInstrumentPanel.cpp:
```diff
@@ -563,6 +559,10 @@ IBFEInstrumentPanel::initializeHierarchyIndependentData(IBFEMethod* ib_method_op
     }
+    
+    // output the nodes of each meter mesh into a .dat file
+    outputNodes();
```
If the point is to handle output outside of the class, should this call to `outputNodes()` still be in the implementation?

cpuelz commented on include/ibamr/IBFEInstrumentPanel.h:
```diff
@@ -100,15 +99,40 @@ class IBFEInstrumentPanel
+    int getNumberOfMeterMeshes() const;
+    
+    /*!
+     * \return get a pointer to the jjth meter mesh
```
I also returned const references for the Mesh and EquationSystems objects.  Is that okay?

cpuelz commented:
I also returned const references for the Mesh and EquationSystems objects. Is that okay?

boyceg commented:
Unfortunately, the interface for the libMesh class `ExodusII_IO` requires a non-`const` reference to a `MeshBase` --- so you would need to return mutable (non-`const`) references to `MeshBase`. I would also return a non-`const` reference to the `EquationSystems` object.

cpuelz commented:
okay... made those changes.
